### PR TITLE
feat(ddm): Implement lenient MRI parsing

### LINKED ISSUE #3108: Align MRI validation between Sentry and Relay
Sentry has its regex to validate MRIs and right now there is a divergence in how Relay validates. The regex definition code in Sentry is here: https://github.com/getsentry/sentry/tree/master/src/sentry/snuba/metrics/naming_layer/mri.py#L71-L71.

This ticket involves the work of aligning the validations in Relay to Sentry. In addition, docs must be updated here: https://develop.sentry.dev/sdk/metrics/#normalization.

### LINKED ISSUE #68157: Implement a more lenient version of MRI parsing in Sentry
Epic: https://github.com/getsentry/team-ingest/issues/304

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR implements a lenient MRI parsing logic since Relay will now perform the escaping and validating logic (https://github.com/getsentry/team-ingest/issues/304).

The idea of having a lenient logic is that we are assuming all data from the system is coming from Relay and this implies that we are treating that as valid. So the parsing just tries to segment the MRI in its components but without checking the actual charset of each component.

Epic: https://github.com/getsentry/team-ingest/issues/304
Closes: https://github.com/getsentry/sentry/issues/68157
Closes: https://github.com/getsentry/relay/issues/3108

### CODE REVIEW

**Comment by ayirr7** [comment]
One initial question; I will return this PR later as well to follow-up with any more in-depth comments. I see that the old `parse_mri` function was one that tries a direct match against the MRI Regex definition in the indexer. Is my understanding correct that, with `parse_mri_lenient`, we are completely bypassing any direct Regex match? 

If that is the case, let's maybe add a comment next to the regex definitions in `mri.py` that those regex are not actively being used or matched against and instead we are using `parse_mri_lenient`

File: src/sentry/snuba/metrics/naming_layer/mapping.py [hunk 1]
```diff
@@ -99,20 +99,17 @@ def is_private_mri(internal_name: TransactionMRI | SessionMRI | str) -> bool:
 
 
 def _extract_name_from_custom_metric_mri(mri: str) -> str | None:
-    match = MRI_SCHEMA_REGEX.match(mri)
-    if match is None:
+    parsed_mri = parse_mri_lenient(mri)
```
Comment (iambriccardo) [inline src/sentry/snuba/metrics/naming_layer/mapping.py:102]: Took the opportunity to remove all of our dependencies from the regex, which now should not be used directly.

**Comment by ayirr7** [comment]
Messaged offline re: additional test cases. 

I think overall this looks good to me. Maybe we can get eyes from someone on the Ingest team as well to ensure that this leniency is ok, and that the indexer can expect to only receive valid MRIs from upstream Relay. With this change, the indexer would no longer be dropping data (so long as it's a structurally correct MRI) so let's make sure upstream expectations match that.

File: src/sentry/snuba/metrics/naming_layer/mri.py [hunk 2]
```diff
@@ -298,11 +307,46 @@ def parse_mri(mri_string: str | None) -> ParsedMRI | None:
...
+            mri_components.append(mri_string[start:i])
+            mri_delimiters.pop(0)
+            start = i + 1
+
+    # Append the substring after the last delimiter. If we have no string after the last delimiter, we will still return
+    # ''.
+    mri_components.append(mri_string[start:])
+
+    # In case we do not find the exact number of components that we expect in the MRI, we will return `None`.
+    if len(mri_components) != 4:
```
Comment (ayirr7) [inline src/sentry/snuba/metrics/naming_layer/mri.py:334]: Did we test this case -- a case where we might not have 4 components due to a missing delimiter?

Comment (iambriccardo) [reply src/sentry/snuba/metrics/naming_layer/mri.py:334]: Oh yeah, I added the edge case of nothing after the `@` but not in the middle. Will add now.

Comment (iambriccardo) [reply src/sentry/snuba/metrics/naming_layer/mri.py:334]: Nvm, I checked again. I did add that test, it's here: https://github.com/getsentry/sentry/blob/6a589b1eafc01321960ba4190cdb131d20fd3276/tests/sentry/snuba/metrics/test_naming_layer.py#L165-L165

Comment (iambriccardo) [reply src/sentry/snuba/metrics/naming_layer/mri.py:334]: And also here: https://github.com/getsentry/sentry/blob/6a589b1eafc01321960ba4190cdb131d20fd3276/tests/sentry/snuba/metrics/test_naming_layer.py#L164-L164

**Comment by ayirr7** [comment]
In terms of possible impact on indexer processing: 

as far as I can tell, the only function this affects for the indexer's message processing is `extract_use_case_id`. The code that calls `extract_use_case_id` is all wrapped in a try-except https://github.com/getsentry/sentry/blob/master/src/sentry/sentry_metrics/consumers/indexer/batch.py#L145-L154, so if there's any exception raised in the `extract_use_case_id` method, we will bubble up an InvalidMessage which gets DLQ'd.

I don't think there's any other method that gets changed in this PR that's also being called within the indexer's message processing but could be something to double check.

File: src/sentry/snuba/metrics/naming_layer/mri.py (hunk 2 above)
Comment (ayirr7) [reply src/sentry/snuba/metrics/naming_layer/mri.py:334]: ok thanks for checking!

**Comment by ayirr7** [comment]
Looks good from my end. Empty unit case is still being actively discussed, so that should be resolved first before merging.

**Review by ayirr7 — APPROVED** [review]

**Review by iker-barriocanal — COMMENTED** [review]
(I don't have a lot of context here)

> we are assuming all data from the system is coming from Relay and this implies that we are treating that as valid

If we make that assumption, why are we doing any validation at all? How flexible should we be?

> I kept the parse_mri function in code since I wanted to still have a function that is stricter in case for any reason we will need in the future

I think we should delete the code that we don't need. If we need it in the future, we can reintroduce it.

File: tests/sentry/snuba/metrics/test_naming_layer.py [hunk 3]
```diff
@@ -104,12 +105,72 @@ def test_invalid_mri_schema_regex(name):
...
+            "c:transactions/measurements.db_calls@none",
+            ParsedMRI("c", "transactions", "measurements.db_calls", "none"),
+        ),
+        (
+            "s:sessions/error@none",
+            ParsedMRI("s", "sessions", "error", "none"),
+        ),
+        (
+            "dist:my_namespace/organizations/v1/my endpoint@{none}",
+            ParsedMRI("dist", "my_namespace", "organizations/v1/my endpoint", "{none}"),
```
[the remark this answers is not in the document]
Comment (iambriccardo) [reply tests/sentry/snuba/metrics/test_naming_layer.py:64]: We should accept it. We are very flexible on purpose.

**Review by jan-auer — APPROVED** [review]

**Pull request merged**